src/idc/zephyr_idc.c

```diff
@@ -120,6 +129,18 @@ int idc_send_msg(struct idc_msg *msg, uint32_t mode)
 	int ret;
 	int idc_send_memcpy_err __unused;
 
+	if (!cpu_is_core_enabled(target_cpu)) {
+		tr_err(&zephyr_idc_tr, "Core %u is down, cannot sent IDC message", target_cpu);
+		return -EACCES;
+	}
+
+	if (zmsg->work_status == WORK_SUBMITTED) {
+		work->sync = 1;
+		ret = k_p4wq_wait(work, K_USEC(IDC_TIMEOUT));
+		if (ret < 0 && zmsg->work_status == WORK_SUBMITTED)
+			return ret;
```

Comment:
have we ever seen this race occurring? Between what and what? Non-blocking operations usually return an error when the action cannot be completed immediately. Wouldn't that be better?
In fact this seems racy too. What if the handler completes between checking for `WORK_SUBMITTED` and calling `k_p4wq_wait()`? Then we'll never be woken up?

Reply:
I can hit this on a MTL with modified multicore tplg. Still trying to find more info on the scenario as I agree this should not happen normally (and it's rare!). I now reworked the wait logic to only rely o p4wq semaphore, so this should be safe against races (and dropped the work_status field).